### PR TITLE
Added rescheduler

### PR DESCRIPTION
**Pull request by Raffo**

### CODE REVIEW

File: cluster/userdata-master.yaml [hunk 1]
```diff
@@ -1134,6 +1134,44 @@ write_files:
             imagePullSecrets:
             - name: pierone.stups.zalan.do
 
+  - path: /srv/kubernetes/manifests/daemonsets/rescheduler.yaml
+    content: |
+      apiVersion: v1
+      kind: Pod
```
Comment (hjacobs) [inline cluster/userdata-master.yaml]: `Pod` but in `daemonsets` folder?

**Comment by hjacobs** [comment]
@Raffo can you give us some details on the "why" and the "how"?

**Comment by hjacobs** [comment]
`matchLabels` is missing I think (depends on `kind`)

File: cluster/userdata-master.yaml [hunk 2]
```diff
@@ -1134,6 +1134,44 @@ write_files:
...
+        namespace: kube-system
+        labels:
+          application: rescheduler
+          version: v0.2.1
+          kubernetes.io/cluster-service: "true"
+          kubernetes.io/name: "Rescheduler"
+      spec:
+        hostNetwork: true
+        containers:
+        - image: gcr.io/google_containers/rescheduler:v0.2.1
```
Comment (hjacobs) [inline cluster/userdata-master.yaml]: We should move it to OSS registry...

File: cluster/userdata-master.yaml (hunk 1 above)
Comment (Raffo) [inline cluster/userdata-master.yaml]: Yeah, that's wrong, will fix it.

**Comment by Raffo** [comment]
Yes: 
Critical pods (our system pods) are guaranteed to be running by the rescheduler. Originally we thought the scheduler included this functionalities, but it does not and we need this additional pod. 

- https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduling.md
- http://kubernetes.io/docs/admin/rescheduler/

File: cluster/userdata-master.yaml (hunk 2 above)
Comment (Raffo) [inline cluster/userdata-master.yaml]: I will rebuild it and push it there.

**Comment by Raffo** [comment]
It will be a `pod` (similar to the scheduler), so no `matchLabels`. Addressed the other comments.

File: cluster/userdata-master.yaml [hunk 3]
```diff
@@ -1134,6 +1134,44 @@ write_files:
...
+          # TODO: Make resource requrements depend on the size of the cluster
+          resources:
+            requests:
+              cpu: 10m
+              memory: 100Mi
+          command:
+          # TODO: split this out into args when we no longer need to pipe stdout to a file #6428
+          - sh
+          - -c
+          - '/rescheduler --running-in-cluster=false 1>>/var/log/rescheduler.log 2>&1'
```
Comment (hjacobs) [inline cluster/userdata-master.yaml]: why do we log into special file? docker logs + log shipping should be good enough...

File: cluster/userdata-master.yaml [hunk 4]
```diff
@@ -1134,6 +1134,44 @@ write_files:
...
+        volumes:
+        - hostPath:
+            path: "/var/log"
+          name: logfile
+
```
Comment (Raffo) [inline cluster/userdata-master.yaml]: I removed it, but it's currently one line everywhere...

**Comment by szuecs** [comment]
:+1:

File: cluster/userdata-master.yaml [hunk 5]
```diff
@@ -1226,6 +1226,34 @@ write_files:
...
+        containers:
+        - image: registry.opensource.zalan.do/teapot/rescheduler:v0.2.1
+          name: rescheduler
+          # TODO: Make resource requrements depend on the size of the cluster
+          resources:
+            requests:
+              cpu: 10m
+              memory: 100Mi
+          command:
+          # TODO: split this out into args when we no longer need to pipe stdout to a file #6428
```
Comment (hjacobs) [inline cluster/userdata-master.yaml]: you did the TODO :smirk:

Comment (hjacobs) [inline cluster/userdata-master.yaml]: partly...

**Comment by hjacobs** [comment]
:+1:

**Comment by Raffo** [comment]
👍

**Pull request merged**